app/zapier/files/actions.js

```diff
@@ -256,11 +266,82 @@ const roomCreate = {
   }
 }
 
+const uploadFile = {
+  key: "uploadFile",
+  noun: "File",
+  display: {
+    label: "Upload File",
+    description: "Upload a file."
+  },
+  operation: {
+    inputFields: [
+      {
+        label: "Room",
+        key: "id",
+        required: true,
+        altersDynamicFields: true,
+        dynamic: "roomCreated.id.title"
+      },
+      {
+        label: "Folder",
+        key: "folderId",
+        dynamic: "folderCreated.id.title"
+      },
+      {
+        label: "URL or File",
+        key: "url",
+        required: true,
+        helpText: "Download file via direct link or hydrate file"
+      }
+    ],
+    /**
+     * @param {ZObject} z
+     * @param {Bundle<SessionAuthenticationData, UploadFileFields>} bundle
+     * @returns {Promise<UploadFileData>}
+     */
+    async perform(z, bundle) {
+      if (!bundle.inputData.folderId && bundle.inputData.id) {
+        bundle.inputData.folderId = bundle.inputData.id
+      }
+      const client = new Client(bundle.authData.baseUrl, z.request)
+      const files = new FilesService(client)
+      const upload = new Upload(z)
+      try {
+        const stashedFile = await upload.stash(bundle.inputData.url)
+        const headers = await upload.headers(stashedFile)
+        const bodySession = {
+          folderId: bundle.inputData.folderId,
+          FileName: headers.fileName,
+          FileSize: headers.fileSize,
+          CreateOn: new Date().toISOString()
+        }
+        const session = await files.createSession(bundle.inputData.folderId, bodySession)
+        const uploadLocation = `${bundle.authData.baseUrl}/ChunkedUploader.ashx?uid=${session.data.id}`
+        const result = await upload.upload(
+          (/** @type {ChunkData} */ chunkData) => files.uploadChunk(chunkData),
+          uploadLocation,
+          stashedFile,
+          headers.fileName,
+          headers.fileSize
+        )
+        return result
+      } catch (error) {
+        if (error instanceof Error) {
+          throw new z.errors.HaltedError(error.message)
+        }
+        throw new z.errors.HaltedError("Unknown error")

```

Comment:
Perhaps it will be better this way.

```js
let message = "Unknown error"
if (error instanceof Error) {
  message = error.message
}
throw new z.errors.HaltedError(message)
```

Reply:
> Perhaps it will be better this way.
> 
> ```js
> let message = "Unknown error"
> if (error instanceof Error) {
>   message = error.message
> }
> throw new z.errors.HaltedError(message)
> ```

Thank you! Rewrote.